### PR TITLE
Drive time

### PR DESCRIPTION
**Pull request by atirabagwan**
drive start time left

### CODE REVIEW

File: app/controllers/drives_controller.rb [hunk 1]
```diff
@@ -0,0 +1,15 @@
+class DrivesController < ApiController
+  def drive_time_left
+    @drive = Drive.find(params[:id])
+
+    drive_time = @drive.start_time.localtime
+    current_time = DateTime.now.localtime
+
+    time_left = drive_time - current_time
+    if time_left.negative?
+      render json: { message: 'drive already started' }
```
Comment (rahulknojha) [inline app/controllers/drives_controller.rb:10]: what will be status of HTTP response here 200 ?

File: app/controllers/drives_controller.rb [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+    else
+      render json: { time_left: time_left }
```
Comment (rahulknojha) [inline app/controllers/drives_controller.rb:12]: Add two methods in ApiController say `render_success` and `render_error`, so that we don't need to repeat render json: '', and http status

Signature of methods will be something like this.

```
  def render_success(data: nil, message: nil)
  end
  
  def render_error(message: nil)
  end
```

**Review by rahulknojha — CHANGES_REQUESTED** [review]
After multiple feedbacks  I can still see following issues.
 - End of new line is missing in one file
 - Still I can see files committed under test folder. 
 
Apart from these. Please add rspec test cases for changes you have made.

File: config/routes.rb [hunk 3]
```diff
@@ -4,6 +4,8 @@
   # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
   resources :statuses, only: [:index]
 
+  get '/drive/:id', to: "drives#drive_time_left"
```
Comment (rahulknojha) [inline config/routes.rb:7]:
```
resources :drives do
     get :drive_time_left, on: :member
end
```

File: app/controllers/candidates_controller.rb [hunk 4]
```diff
@@ -1,17 +1,36 @@
...
+  def candidate_test_time_left
+    @drive_candidate.save if @drive_candidate && @drive_candidate.start_time.nil?
+
+    time_left = (@duration.to_f * 60) - (DateTime.now.localtime - @drive_candidate.start_time.localtime).to_f
+
+    if time_left.negative?
+      render_success(data: { time_left: time_left }, message: I18n.t('test.time_over'), status: 200)
     else
-      render_error(message: I18n.t('not_found.message'), status: 404)
+      render_success(data: { time_left: time_left }, message: I18n.t('test.time_remaining'), status: 200)
```
Comment (rahulknojha) [inline app/controllers/candidates_controller.rb:33]: In above two render_success statement only message is changing so can we make only message condition instead of entire statement. 

```
 
```

File: app/controllers/candidates_controller.rb [hunk 5]
```diff
@@ -22,9 +41,20 @@ def candidate_params
                   :updated_at, :created_by_id, :updated_by_id)
   end
 
-  def find_candidate
-    token = params[:id].to_s
-    @drive_candidate = DrivesCandidate.find_by(token: token)
-    return render_error(message: I18n.t('token_not_found.message'), status: :not_found) if @drive_candidate.blank?
+  def load_drive
+    # byebug
```
Comment (rahulknojha) [inline app/controllers/candidates_controller.rb:45]: what is this ?

File: app/controllers/candidates_controller.rb [hunk 6]
```diff
@@ -22,9 +41,20 @@ def candidate_params
...
+    @drive = Drive.find_by(id: params[:drife_id])
+  end
+
+  def load_duration
+    @duration = @drive.duration if @drive
```
Comment (rahulknojha) [inline app/controllers/candidates_controller.rb:50]: this can be written as

```
@duration = @drive&.duration
```

File: app/controllers/candidates_controller.rb [hunk 7]
```diff
@@ -1,17 +1,36 @@
...
 
   def update
-    if (candidate = @drive_candidate.candidate)
-      if candidate.update(candidate_params)
-        render_success(data: candidate, message: I18n.t('success.message'))
-      else
-        render_error(message: I18n.t('error.message'))
-      end
+    candidate = Candidate.find(params[:id])
+    return unless candidate.update(candidate_params)
```
Comment (rahulknojha) [inline app/controllers/candidates_controller.rb:19]: this has been discussed in earlier PR as well, what it is going to return to the client in case of failure ? 

Self review missing

File: app/controllers/drives_controller.rb [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+    if @time_left_to_start.negative?
+      if @time_left_already_stated.positive?
+        render_success(data: @time_left_already_stated, message: I18n.t('drive.started'),
+                       status: 200)
+      else
+        render_success(data: @time_left_already_stated, message: I18n.t('drive.ended'),
+                       status: 200)
+      end
+    else
+      render_success(data: @time_left_to_start, message: I18n.t('drive.yet_to_start'), status: 200)
```
Comment (rahulknojha) [inline app/controllers/drives_controller.rb:17]: Here as well data and msg is only conditional not entire statement,

File: app/controllers/drives_controller.rb [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+  before_action :load_drive
+  before_action :set_time_data
+
+  def drive_time_left
+    if @time_left_to_start.negative?
+      if @time_left_already_stated.positive?
+        render_success(data: @time_left_already_stated, message: I18n.t('drive.started'),
+                       status: 200)
+      else
+        render_success(data: @time_left_already_stated, message: I18n.t('drive.ended'),
```
Comment (rahulknojha) [inline app/controllers/drives_controller.rb:13]: do we need to send status explicitly I guess it should be default status

File: app/controllers/drives_controller.rb [hunk 10]
```diff
@@ -0,0 +1,34 @@
+# frozen_string_literal: true
+
+class DrivesController < ApiController
+  before_action :load_drive
+  before_action :set_time_data
+
+  def drive_time_left
+    if @time_left_to_start.negative?
+      if @time_left_already_stated.positive?
```
Comment (rahulknojha) [inline app/controllers/drives_controller.rb:9]: started ?

File: app/controllers/drives_controller.rb [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+  private
+
+  def load_drive
+    @drive = Drive.find(params[:id])
+  end
+
+  def set_time_data
+    @drive_start_time = @drive.start_time.localtime
+    @drive_end_time = @drive.end_time.localtime
+    @current_time = DateTime.now.localtime
```
Comment (rahulknojha) [inline app/controllers/drives_controller.rb:30]: do we need all above three variables as instance variable ?

File: config/database.yml [hunk 12]
```diff
@@ -20,6 +20,8 @@ default: &default
   # For details on connection pooling, see Rails configuration guide
   # http://guides.rubyonrails.org/configuring.html#database-pooling
   pool: <%= ENV.fetch("RAILS_MAX_THREADS") { 5 } %>
+  username: postgres
+  password:
```
Comment (rahulknojha) [inline config/database.yml:24]: avoid un intentional commit.

File: spec/request/candidates_spec.rb [hunk 13]
```diff
@@ -3,32 +3,60 @@
...
+        id: 0,
+        first_name: Faker::Name.name
+      }
+      put :update, params: params
+
+      expect(response).to have_http_status(:not_found)
+    end
+  end
+
+  describe 'candidate test time' do
```
Comment (rahulknojha) [inline spec/request/candidates_spec.rb:37]: it should mention action name

File: spec/request/candidates_spec.rb [hunk 14]
```diff
@@ -3,32 +3,60 @@
...
+
+    it 'raises error exception if particular candidate is not found' do
+      params = {
+        id: 0,
+        first_name: Faker::Name.name
+      }
+      put :update, params: params
+
+      expect(response).to have_http_status(:not_found)
+    end
```
Comment (rahulknojha) [inline spec/request/candidates_spec.rb:34]: add test cases for negative scenario like first name is not present.

File: spec/request/drives_spec.rb [hunk 15]
```diff
@@ -0,0 +1,62 @@
+# frozen_string_literal: true
```
Comment (rahulknojha) [inline spec/request/drives_spec.rb:1]: why we are writing request specs here ?, we are using controller spec right then move it to controller folder

**Review by rahulknojha — CHANGES_REQUESTED** [review]

File: spec/request/drives_spec.rb (hunk 15 above)
Comment (varungadde99) [reply spec/request/drives_spec.rb:1]: Moved the _specs.rb files to app/controllers folder post discussion.

File: config/database.yml (hunk 12 above)
Comment (varungadde99) [reply config/database.yml:24]: file post merge added to .gitignore.

File: app/controllers/candidates_controller.rb (hunk 7 above)
Comment (varungadde99) [reply app/controllers/candidates_controller.rb:19]: Added render_error in the else scenario where due to wrong params the update can fail.

File: spec/request/candidates_spec.rb (hunk 14 above)
Comment (varungadde99) [reply spec/request/candidates_spec.rb:34]: Pending for discussion.

File: config/database.yml (hunk 12 above)
Comment (varungadde99) [reply config/database.yml:24]: Sir still the file is added to .gitignore, still its not resolving this error.

**Review by rahulknojha — APPROVED** [review]
